spyder/plugins/ipythonconsole/widgets/debugging.py

```diff
@@ -28,27 +28,57 @@ class DebuggingWidget(RichJupyterWidget):
     Spyder
     """
 
+    def __init__(self, *args, **kwargs):
+        super(DebuggingWidget, self).__init__(*args, **kwargs)
+        self._previous_prompt = None
+        self._input_queue = []
+        self._input_ready = False
+
+    def monkeypatch_kernel_client(self, client):
+        """Monkeypatches the kernel client to monitor input calls."""
+        if client:
+            old_input = client.input
+
+            def monkeypatched_input(string):
+                """If input is not ready, save it in a queue."""
+                if self._input_ready:
+                    self._input_ready = False
+                    return old_input(string)
+                elif self.is_debugging():
+                    self._input_queue.append(string)
+
+            client.input = monkeypatched_input
+        return client
+
+    def register_message_handler(self, spyder_kernel_comm):
```

Comment:
It's very confusing to have several `register_message_handler` methods in the different widgets that receive `spyder_kernel_comm` as an argument, when a) `spyder_kernel_comm` is an attribute of `ShellWidget` and b) these widgets are merely a code trick to not have a gigantic `ShellWidget`.

Therefore, I think we should unify all these `register_message_handler` methods in `ShellWidget` and call it only there (on it's `__init__` method, as you're doing it already).

Reply:
However, the handler definitions can keep living in their respective widgets (e.g. `get_spyder_breakpoints` here and `set_pdb_state` in `NamepaceBrowserWidget`).